pyro/contrib/forecast/util.py

```diff
@@ -0,0 +1,228 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from functools import singledispatch
+
+import torch
+
+import pyro.distributions as dist
+from pyro.poutine.messenger import Messenger
+from pyro.poutine.util import site_is_subsample
+
+
+class PrefixReplayMessenger(Messenger):
+    """
+    EXPERIMENTAL Given a trace of training data, replay a model with batched
+    sites extended to include both training and forecast time, using the guide
```

Comment:
Oops good catch. This should not assume dim=-2. I think this will need a two part solution:
1. inside the `time_plate` set `dim=-1 - msg["fn"].event_dim`
2. outside the `time_plate` we still need to support [reparametrizer logic](https://github.com/pyro-ppl/pyro/blob/ce76668176b6df2d46f5d3276a232b913740c0d8/pyro/infer/reparam/hmm.py#L80), so we can search to find which dim differs.

I'll push a fix soon.

Reply:
OK, I've changed to simply search for whatever dimension differs, ignoring the `time_plate`. I've also added a new test case where `dim=-` that would have broken before your comment.